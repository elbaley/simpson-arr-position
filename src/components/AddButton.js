import {StyleSheet, Text, View, TouchableOpacity} from 'react-native';
import React from 'react';

const AddButton = ({onPress}) => {
  return (
    <TouchableOpacity style={styles.container} onPress={onPress}>
      <Text style={styles.buttonText}>Add</Text>
    </TouchableOpacity>
  );
};

export default AddButton;

const styles = StyleSheet.create({
  container: {
    backgroundColor: 'blue',
    justifyContent: 'center',
    alignItems: 'center',
    alignSelf: 'center',
    margin: 20,
    padding: 20,
    paddingHorizontal: 30,
    borderRadius: 20,
  },
  buttonText: {
    color: '#ffffff',
  },
});
