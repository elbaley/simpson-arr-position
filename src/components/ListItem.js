import {
  StyleSheet,
  Text,
  View,
  TouchableOpacity,
  Image,
  Alert,
} from 'react-native';
import React, {useContext} from 'react';
import {useNavigation} from '@react-navigation/native';
import {colors} from '../styles/theme';
import {AppContext} from '../context/AppContext';

export const ListItem = ({item, index, removeListItem}) => {
  const {data, setData} = useContext(AppContext);
  const canMoveUp = index === 0 ? false : true;
  const canMoveDown = index === data.length - 1 ? false : true;

  const handleMove = direction => {
    let arr = [...data];
    const fromIndex = arr.findIndex(d => d.id === item.id);
    let toIndex = -1;
    switch (direction) {
      case 'up':
        toIndex = fromIndex - 1;
        break;
      case 'down':
        toIndex = fromIndex + 1;
        break;
      default:
    }

    if (toIndex >= arr.length || toIndex < 0) return;

    const element = arr.splice(fromIndex, 1)[0];
    arr.splice(toIndex, 0, element);
    setData(arr);
  };

  const alertBeforeDelete = () => {
    Alert.alert('You are deleting person', 'Are you sure', [
      {
        text: 'Cancel',
        onPress: () => console.log('Cancel Pressed'),
        style: 'cancel',
      },
      {text: 'OK', onPress: () => removeListItem({item})},
    ]);
  };

  const navigation = useNavigation();
  return (
    <TouchableOpacity
      style={styles.renderItemContainer}
      onPress={() => navigation.navigate('Detail', {item: item})}>
      <View style={{flexDirection: 'row', alignItems: 'center'}}>
        <Text>{index + 1} </Text>
        <Image
          source={{uri: item.avatar}}
          style={styles.imageStyle}
          resizeMode={'center'}
        />
        <Text>{item.name}</Text>
      </View>
      <View style={{flexDirection: 'row'}}>
        {canMoveUp && (
          <TouchableOpacity onPress={() => handleMove('up')}>
            <Text style={styles.renderItemButtonText}>↑</Text>
          </TouchableOpacity>
        )}

        {canMoveDown && (
          <TouchableOpacity onPress={() => handleMove('down')}>
            <Text style={styles.renderItemButtonText}>↓</Text>
          </TouchableOpacity>
        )}
        <TouchableOpacity onPress={() => alertBeforeDelete()}>
          <Text style={styles.renderItemButtonText}>X</Text>
        </TouchableOpacity>
      </View>
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  renderItemContainer: {
    backgroundColor: colors.white,
    flexDirection: 'row',
    padding: 10,
    marginTop: 1,
    justifyContent: 'space-between',
    alignItems: 'center',
  },
  renderItemButtonText: {
    fontSize: 24,
    color: '#000000',
    fontWeight: '600',
    marginHorizontal: 10,
  },
  imageStyle: {
    width: 40,
    height: 40,
  },
});
