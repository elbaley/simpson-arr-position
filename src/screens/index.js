export {default as ListScreen} from './ListScreen';
export {default as DetailScreen} from './DetailScreen';
export {default as AddScreen} from './AddScreen';