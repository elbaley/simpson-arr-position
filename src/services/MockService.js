import axios, {AxiosResponse} from 'axios';
import React, {useContext} from 'react';
import { HelperFunctions } from '../helpers/ServiceHelpers';

const url = HelperFunctions.GetSimpsonsServiceUrl();

export default MockService = () => {

  const getData = async () => {
    try {
      let res = await axios.get(url);

      return res;
    } catch (error) {
      return error;
    }
  };

  return {getData};
};
