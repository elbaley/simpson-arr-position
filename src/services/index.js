export {default as MockService} from '../services/MockService';