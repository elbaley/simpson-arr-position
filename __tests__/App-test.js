import 'react-native';
import React from 'react';
import {render} from '@testing-library/react-native';
import {ListItem} from '../src/components/ListItem';
import axios from 'axios';
import {MockService} from '../src/services';

const mockedNavigate = jest.fn();
jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({navigate: mockedNavigate}),
}));

jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useNavigation: () => {
    return mockedNavigate;
  },
}));

it('list item renders correctly', () => {
  const fakeItems = {
    item: {
      name: 'Homer Simpsons',
    },
    index: 1,
  };
  const {getByText, toJSON} = render(<ListItem items={fakeItems} />);
  expect(getByText('Homer Simpsons')).toBeTruthy();
  expect(toJSON()).toMatchSnapshot();
});

jest.mock('axios');
const {getData} = MockService();

describe('Get simpsons', () => {
  describe('when API call is successful', () => {
    it('should return users list', async () => {
      // given

      const mockData = {
        item: {
          name: 'Homer Simpsons',
          avatar:
            'https://static.wikia.nocookie.net/simpsons/images/b/bd/Homer_Simpson.png/revision/latest/scale-to-width-down/300?cb=20201222215437',
          job: 'Nuclear Safety Inspector',
          description:
            "Homer Jay Simpson (born May 12, 1956) is the main protagonist and one of the five main characters of The Simpsons series (or show). He is the spouse of Marge Simpson and father of Bart, Lisa and Maggie Simpson. Homer is overweight (said to be ~240 pounds), lazy, and often ignorant to the world around him. Although Homer has many flaws, he has shown to have great caring, love, and even bravery to those he cares about and, sometimes, even others he doesn't. He also serves as the main protagonist of the The Simpsons Movie. He is 39 years old and was born in 1956.",
          id: '14',
        },
      };

      axios.get.mockResolvedValueOnce(mockData);
      // when
      const result = await getData();
      //console.log(result);
      // then
      expect(axios.get).toHaveBeenCalledWith(
        'https://5fc9346b2af77700165ae514.mockapi.io/simpsons',
      );
      expect(result).toEqual(mockData);
    });
  });
});
