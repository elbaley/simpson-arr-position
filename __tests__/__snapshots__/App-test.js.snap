// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`list item renders correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#FFFFFF",
      "flexDirection": "row",
      "justifyContent": "space-between",
      "marginTop": 1,
      "opacity": 1,
      "padding": 10,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
      }
    }
  >
    <Text>
      2
       
    </Text>
    <Image
      resizeMode="center"
      source={
        {
          "uri": undefined,
        }
      }
      style={
        {
          "height": 40,
          "width": 40,
        }
      }
    />
    <Text>
      Homer Simpsons
    </Text>
  </View>
  <View
    style={
      {
        "flexDirection": "row",
      }
    }
  >
    <View
      accessible={true}
      collapsable={false}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          {
            "color": "#000000",
            "fontSize": 24,
            "fontWeight": "600",
            "marginHorizontal": 10,
          }
        }
      >
        ↑
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          {
            "color": "#000000",
            "fontSize": 24,
            "fontWeight": "600",
            "marginHorizontal": 10,
          }
        }
      >
        ↓
      </Text>
    </View>
    <View
      accessible={true}
      collapsable={false}
      focusable={false}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          {
            "color": "#000000",
            "fontSize": 24,
            "fontWeight": "600",
            "marginHorizontal": 10,
          }
        }
      >
        X
      </Text>
    </View>
  </View>
</View>
`;
